Paper: Deep and Ensemble Learning to Win the Army RCO AI Signal Classification Challenge (#473)
scipy-conference/scipy_proceedings: papers/aero_nn_vila/aero_nn_vila.rst

:author: Eugene Grayver
:email: eugene.grayver@aero.org
:institution: The Aerospace Corporation

---------------------------------------------------------------------------------
Deep and Ensemble Learning to Win the Army RCO AI Signal Classification Challenge
---------------------------------------------------------------------------------

.. class:: abstract

Automatic modulation classification is a challenging problem with multiple
applications including cognitive radio and signals intelligence. Most of the
existing efforts to solve this problem are only applicable when the signal to
noise ratio (SNR) is high and/or long observations of the signal are available.
Recent work has focused on applying shallow and deep machine learning (ML) to
this problem. In this paper, we present an exploration of such deep learning and ensemble
learning techniques that was used to win the Army Rapid Capability Office (RCO)
2018 Signal Classification Challenge. An expert feature extraction and shallow
learning approach is discussed in a simultaneous publication. We evaluated multiple
state-of-the-art deep learning network architectures and adapted them to work in the

> RF signal domain instead of the image/computer vision domain. The best deep learning methods

Reviewer: ```suggestion
RF signal domain instead of the image/computer-vision domain. The best deep learning methods
```

were merged with the best expert feature extraction and shallow learning

> methods using ensemble learning . Finally, the ensemble classifier was calibrated to obtain marginal gains.

Reviewer: ```suggestion
methods using ensemble learning. Finally, the ensemble classifier was calibrated to obtain additional marginal gains.
```

The methods discussed are capable of correctly classifying waveforms at -10 dB SNR
with over 63% accuracy and signals at +10 dB SNR with over 95% accuracy from an Army RCO provided training set.

.. class:: keywords

   modulation classification, neural networks, deep learning, machine learning, ensemble learning,
   wireless communications, signals intelligence, probability calibration

Introduction
------------

All conventional communications systems are designed with the assumption that
the transmitter and receiver are cooperative and have full knowledge of the
waveform being exchanged. However, there are scenarios where the receiver does
not know what waveform (i.e. modulation, coding, etc.) has been transmitted.
Classical examples include cognitive radio network (i.e. a new terminal enters a
network and needs to figure out what waveform is being used), and signals

> intelligence (i.e. interception of adversary’s communications). The problem of

Reviewer: ```suggestion
intelligence (i.e. interception of an adversary’s communications). The problem of
```

waveform classifications, or more narrowly, modulation recognition has been
studied for decades [Aisbett]_. Given the implication of SIGINT [#]_ applications
before cognitive radio, much of the work had not been published. Key early work
is done by Azzouz & Nandi [Nandi1]_ [Nandi2]_ [Azz1]_ [Azz2]_.

The fundamental approach taken by most authors has been to find data reduction
functions that accentuate the differences between different waveforms. These

> functions are applied to input samples and decision is made by comparing the

Reviewer: ```suggestion
functions are applied to input samples and a decision is made by comparing the
```

values against a set of multi-dimensional thresholds. Determining the threshold
values by hand becomes impractical as the number of clusters and/or functions
grows. The idea to apply neural networks to help make these decisions has been
around for decades [Azz2]_. However, it is only recently that our
understanding of machine learning combined with enormous increase in
computational resources has enabled us to use ML techniques to solve this problem.

.. [#] Signals Intelligence

Challenge Description
---------------------

.. figure:: data-flow-full.pdf
    :align: center
    :figclass: w
    :scale: 60%

>     Data flow through the classification pipeline. Light-blue denote the many variable parameters available.

Reviewer: ```suggestion
    Data flow through the classification pipeline. The many variable parameters available are denoted in light blue.
```

    :label:`data-flow`

The Army Rapid Capability Office is seeking innovative approaches to leverage
artificial intelligence (AI) to conduct blind radio frequency signal analysis.
To this end, they published a labeled modulation classification dataset and
created a competition [Army]_ to properly classify a pair of unlabeled test
sets. This paper details the efforts of The Aerospace Corporation’s Team
Platypus, the authors of this paper, to build a modulation classification system via deep learning and ensemble
learning. In this context, deep refers to the fact that the ML classifier will use the raw IQ [#]_
data, instead of expertly engineered features. The winning submission from Team
Platypus utilized a combination of these deep classifiers and shallow learning classifiers
built on expert features which are described in a simultaneous companion publication.

.. [#] In-Phase & Quadrature

The training dataset [Mitre]_ consists of 4.32 million signals each of which

> containing 1024 complex (IQ) points and a label indicating the modulation type

Reviewer: ```suggestion
contain 1024 complex (IQ) points and a label indicating the modulation type
```

modulations (including a noise class), with AWGN at six different
signal-to-noise ratios (-10, -6, -2, +2, +6, or +10 dB). The complete dataset
included 30,000 rows for each modulation and SNR configuration. Sample rate is
selected from a set (200, 500, 1000, or 2000 ksps), and symbol rate is selected
from a set (4, 8, 16, or 32 samples per symbol). Neither of the rate parameters
is included in the label.

The competition consisted of assigning a likelihood score to each of the 24
possible modulation classes for each of the 100,000 rows in a pair of unlabeled
test sets.

Classifier performance is evaluated via a pre-defined equation based on the
well-known log loss metric. The traditional log loss equations:

.. math::
    :label: logloss

    logloss = -\dfrac{1}{N}\sum ^{N}_{i=1}\sum ^{M}_{j=1}y_{ij}\log p_{ij}

Where N is the number of instances in the test set, M is the number of
modulation class labels (24), :math:`y_{ij}` is 1 if test instance :math:`i`
belongs to class :math:`j` and 0 otherwise, :math:`p_{ij}` is the predicted
probability that observation :math:`i` belongs in class :math:`j`. The competition score,

> which we will refer to it as simply the score in the remainder of this paper,

Reviewer: ```suggestion
which we will refer to as simply the score in the remainder of this paper,
```

was defined per [Mitre]_ as follows:

.. math::
    :label: score

    score = \dfrac {1}{1+logloss}

Notes:

* A uniform probability estimate would yield a score of 23.935

* To get a perfect 100 score participants would need to be both 100% correct and 100% confident of every estimation.

We will also use a more standard :math:`F_1` metric for each modulation is used.
This is an excellent measurement of classifier performance since it uses both
recall :math:`r` and precision :math:`p`, which better account for false negatives
and false positives:

.. math::
    :label: recall

    r = \dfrac{\sum {true\ positive}}{\sum {false\ negative}+\sum {true\ positive}}

.. math::
    :label: precision

    p = \dfrac{\sum {true\ positive}}{\sum {false\ positive}+\sum {true\ positive}}

.. math::
    :label: f1

    F_1 = \dfrac {2}{\frac {1}{r}+\frac {1}{p}}

Approach
--------

Team Platypus' approach to solve this modulation classification problem is to

> combine deep neural networks and a shallow learning classifiers leveraging

Reviewer: ```suggestion
combine deep neural networks and shallow learning classifiers leveraging
```

custom engineering features. Both of these are supervised machine learning
systems.

Figure :ref:`data-flow` shows the general flow of data through our winning
system. The labeled training data is split into training,
cross-validation, and testing using a 70%-15%-15% split. When using neural
networks, the cross-validation set is used to prevent classifier
overfitting. Using the Army RCO score metric, the final version of this
system scored 76.422. This equates to a cross-validation log loss of 0.308. The
output of each step is written to large cache files to enable quick evaluation
of new ideas and integration into the next processing pipeline.

Classification Strategy & Scores
--------------------------------

There were two unlabeled sets released to competitors. Estimates generated for
the first set using our deep neural network estimator resulted in very low and
inconsistent scores. It was apparent that the data was very unlike the training
data initially provided. Team Platypus estimates that only half of the first

> unlabeled set was like the training set. Our solutions for this datasets relied

Reviewer: ```suggestion
unlabeled set was like the training set. Our solutions for these datasets relied
```

The final step in the pipeline presented in Figure :ref:`data-flow` is calibration. Probability calibration consists on
modifying the final probabilities without changing the class that corresponds to the highest probability.
It uses the 15% cross-validation data to shape the output probabilities to increase the score.

In order to calibrate our merging neural network we used a modification of the temperature scaling approach
proposed in [Guo]_. The temperature scaling approach finds the optimal temperature scalar
to divide the the output logits by, that minimizes the log loss on the cross-validation dataset.
We extended this method by finding the separate optimal temperature scalars for each predicted modulation type using the
cross-validation data. Temperature scaling consistently increased the score of neural nets from 0.3 to 0.6 points.

Calibration of the scikit-learn merging classifiers consisted on using the CalibrateClassifierCV class in
scikit-learn [SKCal]_. This class implements two different approaches for performing calibration: a parametric approach based
on Platt's sigmoid model and a non-parametric approach based on isotonic regression. Our best results were

> achieved with the isotonic approach which were always between 0.1 to 0.9 points better than the pre-calibrated

Reviewer: ```suggestion
achieved with the isotonic approach which were always between 0.1 to 0.9 points better than the uncalibrated
```

?
